Suggestion: Consideration of a scale factor

bjreisman opened the pull request:
Thanks for your work on this package! I find it very useful for preparing figures which I later want to edit in a vector graphics program. One feature I liked about the previous version was the raster.width and raster.height arguments which let me put a 'big figure' in a small plot. This is especially useful with geom_point as otherwise the `size` argument doesn't really work and the dots can get quiet big. For my own use, I forked a copy of the package and added a `scale` argument to `rasterise` which adjusts raster dimensions by multiplying the height and width.  By default, `scale = 1` which preserves the current behavior. I've only really tested it with geom_point but I can't see a reason it would be geom specific. 

Would you consider adding this to the main package? I've included an example below. 
```{r}
library(ggplot2)
library(ggrastr)

plot <- ggplot(diamonds, aes(carat, price, colour = cut))
plot + rasterise(geom_point(), dpi = 300, scale = 1)
```
![image](https://user-images.githubusercontent.com/22528952/109016105-6e637980-767b-11eb-8dc7-59909ef9bda2.png)
```{r}
plot + rasterise(geom_point(), dpi = 300, scale = 2)
```
![image](https://user-images.githubusercontent.com/22528952/109016253-96eb7380-767b-11eb-8ae8-240e984b87bd.png)

```{r}
plot + rasterise(geom_point(), dpi = 300, scale = 0.5)
```
![image](https://user-images.githubusercontent.com/22528952/109016323-b1255180-767b-11eb-9004-376b5ca87192.png)

bjreisman committed:
Update rasterise.R

added scale parameter to adjust width and height

bjreisman committed:
Update rasterise.R

bjreisman committed:
Update rasterise.R

attempt to add scale parameter

evanbiederstedt commented:
Hi @bjreisman 

Thank you for the PR! 

This seems pretty reasonable to me. What are your thoughts @VPetukhov?

evanbiederstedt committed:
edit travis config, install Cairo

evanbiederstedt commented:
Hi @bjreisman 

BTW, there's an error here when using your changes which I haven't had time to look into. The error is here in the vignettes

```
library(ggplot2)
library(ggrastr)
points_num <- 500000
df <- data.frame(x=as.factor(1:points_num %% 2), y=log(abs(rcauchy(points_num))))
gg <- ggplot(df, aes(x=x, y=y)) + scale_color_discrete(guide=FALSE)
gg_box_vec <- gg + geom_boxplot_jitter(outlier.size=0.1, outlier.jitter.width=0.3, outlier.alpha=0.5)
print(gg_box_vec)
```

Currently the Travis builds give `Failed to create Cairo backend!`. We possible need to use `cairoDevice::Cairo()` , cf https://stat.ethz.ch/pipermail/r-help/2010-April/236785.html

If you have some, please have a look. We'll see who gets to the fix first :)

Best, Evan

bjreisman commented:
Thanks all! I'll try to investigate this as well. I tried running that block of code in my r session and it ran without issue, but I'll see if I can reproduce it. 
-Ben

evanbiederstedt commented:
Hi @bjreisman 

Thanks, I appreciate it---I still haven't had time to take a closer look. 

The Travis logs with the error are here: https://travis-ci.com/github/vpetukhov/ggrastr/builds/218111822
I also ran into an issue locally on my Mac OS.

bjreisman committed:
moved height/width rescaling outside of rasterGrob

bjreisman committed:
added scale parameter to geom_boxplot_jitter

bjreisman committed:
updated documentation to include the scale parameter

bjreisman committed:
added scale parameter to all geoms, updated docs

bjreisman committed:
fixed geom_tile_rast default scale

bjreisman commented:
I ran it again on my ubuntu 20.04 box and was able to reproduce it, I believe the newest commits have addressed any issues. I've added the scale parameter to each of the geoms and reran the examples - all appear to work as expected. I've also updated the docs to reflect the new argument. Thanks for your consideration!

evanbiederstedt commented:
Thanks @bjreisman! You beat me to a fix :)

I changed base to the branch `develop`---I hadn't noticed you were working against `master` previously. Please double check nothing has been broken. 

> #' @param scale A numeric of length one, setting the scaling factor (default=1)

It might be useful to detail that it's a multiplicative scaling factors, i.e. scale=1 is unchanged, scale>1 is larger, <1 is smaller.

This along with the other PR can go into version 0.2.3. I'll update the vignettes accordingly.

evanbiederstedt commented:
Hi @bjreisman 

I'm a bit confused. You have a multiplicative factor here:

https://github.com/bjreisman/ggrastr/blob/master/R/rasterise.R#L89-L91

```  
# Rescale height and width
width <- width * scale
height <- height * scale
```

and you divide here:

```
  # Forward raster grob
  grid::rasterGrob(
    cap, x = 0.5, y = 0.5,
    height = unit(height / scale, "inch"),
    width = unit(width / scale, "inch"),
    default.units = "npc",
    just = "center"
  )
  ```

bjreisman commented:
Good question. At first I just tried the multiplication step. That rendered it as a larger plot, but it was only the center of the plot with the rest of the plot cropped out. I tried adding the division step right at the end to resize the plot after it had been rasterized and it seemed to create the desired behavior. I realize this is a bit hacky; there may be a better way to do this but it seemed to work for both large and small factors.
```{r}
## without dividing at the end. 
plot <- ggplot(diamonds, aes(carat, price, colour = cut))
plot + rasterise(geom_point(), dpi = 300, scale = 2)

```
![image](https://user-images.githubusercontent.com/22528952/109356441-55f38a80-7846-11eb-91e0-a7795421ae6d.png)

evanbiederstedt commented:
@bjreisman  Ah, I think I see what's going on now. Thanks. 

I think it makes sense to merge in this PR to `develop` first: https://github.com/VPetukhov/ggrastr/pull/26

We'll have to deal with merge conflicts, but I think this should be fine.

bjreisman committed:
imporved docs for `scale`, required scale > 0.

evanbiederstedt committed:
Merge branch 'develop' into master

evanbiederstedt committed:
check for scale=0

evanbiederstedt commented:
Hi @bjreisman 

I just wanted to clarify something about your implementation: when `scale` is larger than 1, the points appear smaller; when `scale` is smaller than 1, the points appear smaller. 

What is the rationale for not having the opposite behavior?

That is, `scale=1` means unchanged. 

![plot1](https://user-images.githubusercontent.com/6968193/109373257-b94bdf80-787b-11eb-976b-c9d0c468e4c1.png)

`scale=2` means that the points are twice the size:

![plot2](https://user-images.githubusercontent.com/6968193/109373260-c36dde00-787b-11eb-8855-4426f0e0f8bb.png)

`scale=0.5` means that the points are half the size:

![plot3](https://user-images.githubusercontent.com/6968193/109373269-d54f8100-787b-11eb-952b-4e846e4d3fa7.png)

What's the intuition behind the opposite scaling? 

(Accidentally closed PR; re-opened, apologies)

evanbiederstedt commented:
Naturally, I implemented the above by diving first, and then multiplying, i.e. first this:

```
width <- width / scale
height <- height / scale
```
then:

```
# Forward raster grob
grid::rasterGrob(
  cap, x = 0.5, y = 0.5,
  height = unit(height * scale, "inch"),
  width = unit(width * scale, "inch"),
  default.units = "npc",
  just = "center"
)
```

This scheme currently feels more intuitive to me, but I could be missing something.

bjreisman commented:
I don’t have a strong preference either way. When I wanted to achieve this behavior with previous versions of the package I increased the ‘raster.width’ and ‘raster.height’ so I tried to reproduce that behavior in my implementation. In the rasterise example I added to the docs I briefly explained it as rendering the plot as if it was ‘larger’ in the same area. For those who didn’t use the previous version, I concur that ‘larger or smaller dots’ might be more intuitive.

evanbiederstedt commented:
Hmmm. I currently think the logic that "scale=1 is normal, scale=2 is twice the dot size, scale=0.5 is half the dot size" makes more sense. @VPetukhov, please disagree. 

I'll push those changes to the PR, and modify the CHANGELOG/roxygen2.

evanbiederstedt committed:
scaling, update roxygen2 docs

evanbiederstedt committed:
added to changelog

evanbiederstedt reviewed (approved):
LGTM
